aya/src/maps/mod.rs

```diff
@@ -511,28 +629,73 @@ impl MapData {
     pub fn from_fd(fd: OwnedFd) -> Result<Self, MapError> {
         let info = bpf_map_get_info_by_fd(fd.as_fd())?;
 
+        let name_bytes = bytes_of_c_char(&info.name);
         Ok(Self {
             obj: parse_map_info(info, PinningType::None),
             fd: fd.into_raw_fd(),
-            pinned: false,
+            name: std::str::from_utf8(name_bytes)
+                .map_err(|_| MapError::ParseNameError)?
+                .to_string(),
+            paths: HashSet::new(),
         })
     }
 
-    pub(crate) fn pin<P: AsRef<Path>>(&mut self, name: &str, path: P) -> Result<(), PinError> {
+    pub(crate) fn pin<P: AsRef<Path>>(&mut self, path: P) -> Result<(), PinError> {
         use std::os::unix::ffi::OsStrExt as _;
 
-        let Self { fd, pinned, obj: _ } = self;
-        if *pinned {
-            return Err(PinError::AlreadyPinned { name: name.into() });
+        let Self {
+            fd,
+            paths,
+            obj: _,
+            name,
+        } = self;
+        if paths.contains(&path.as_ref().to_path_buf()) {
+            return Err(PinError::AlreadyPinned {
+                name: name.to_string(),
+            });
         }
-        let path = path.as_ref().join(name);
-        let path_string = CString::new(path.as_os_str().as_bytes())
-            .map_err(|error| PinError::InvalidPinPath { path, error })?;
+        let path_string = CString::new(path.as_ref().as_os_str().as_bytes()).map_err(|error| {
+            PinError::InvalidPinPath {
+                path: path.as_ref().to_path_buf(),
+                error,
+            }
+        })?;
+        debug!(
+            "Attempting to pin map {name} at {}",
+            path.as_ref().display()
+        );
         bpf_pin_object(*fd, &path_string).map_err(|(_, io_error)| SyscallError {
             call: "BPF_OBJ_PIN",
             io_error,
         })?;
-        *pinned = true;
+        paths.insert(path.as_ref().to_path_buf());
+        Ok(())
+    }
+
+    pub(crate) fn unpin(&mut self) -> Result<(), PinError> {
```

Comment:
Per comment above, this needs to take a `Path`.
We shouldn't assume that all pins we know about need unpinning.